cmake/CeleritasAddTest.cmake

```
@@ -308,12 +308,49 @@ function(celeritas_add_test SOURCE_FILE)
 
     # Create an executable and link libraries against it
     add_executable(${_TARGET} "${SOURCE_FILE}" ${PARSE_SOURCES})
+    set_target_properties(${_TARGET} PROPERTIES
+      CUDA_SEPARABLE_COMPILATION ON
+      CUDA_RUNTIME_LIBRARY Shared
+    )
+
+    if (BUILD_SHARED_LIBS AND CELERITAS_USE_CUDA)
+      string(FIND "${SOURCE_FILE} ${PARSE_SOURCES}" .cu iscudalinked)
```

This is an approximation (it would match `foo.cuda.cc`) but good enough for now...

---

CMake internally does a similar match (i.e. it needs to decide whether to compile the file with gcc, g++ or nvcc), do you know if CMake has the list of "cuda" extension it recognized available?

---

I think the extension mapping happens inside of the CMake C++ code in a loop over the individual file names, where it robustly extracts the extension and compares against a hard-coded list:
```c++
    setupExts(this->CLikeSourceFileExtensions,
              { "c", "C", "c++", "cc", "cpp", "cxx", "cu", "mpp", "m", "M",
                "mm", "ixx", "cppm" });
    setupExts(this->HeaderFileExtensions,
              { "h", "hh", "h++", "hm", "hpp", "hxx", "in", "txx" });
    setupExts(this->CudaFileExtensions, { "cu" });
    setupExts(this->FortranFileExtensions,
              { "f", "F", "for", "f77", "f90", "f95", "f03" });
    setupExts(this->ISPCFileExtensions, { "ispc" });
```

This is different from concatenating two lists with a space and searching for an extension. The equivalent would be:
```suggestion
      set(_has_cuda FALSE)
      foreach(_file IN LISTS SOURCE_FILE PARSE_SOURCES)
         get_filename_component(_ext "${file}" LAST_EXT)
         if(_ext STREQUAL ".cu")
           set(_has_cuda TRUE)
           break()
         endif()
       endforeach()
```

---

Fair enough.  I misunderstood the direction of your request (i.e. was 'need less matches', I read 'need more matches' :) )

---

Aha, gotcha. And again, I'm fine as-is since it's pretty limited scope, though we should be prepared for some unexpected CUDA linkage sometime in the future ;)

---

Could the [`LANGUAGE`](https://cmake.org/cmake/help/latest/prop_sf/LANGUAGE.html) source file property property be used here (but note the policy change!). That should be set once the file is a added to a target, or if the user has explicitly set it. Won't prevent real corner cases, but might help to sanity check things like the `.cuda.cc` one.

---

I had tried `LINKER_LANGUAGE` but, when implicit, it set too late.  I will try `LANGUAGE`.

---

Based on the documentation it sounds like that property might be "write-only" rather than something that cmake sets under the hood by default based on source extensions... I'll check the cmake source to see if this is in fact the case.

---

oh nope actually it looks like CMake *will* try to figure out the property based on the file extension, even if not set. So yes, that's a much better way than my suggestion :)

---

However the LANGUAGE is (somehow) not always set, so now I use both.